app/fetch-did-web-doc.py

```diff
@@ -119,14 +123,129 @@ def request(url, data=None, headers=None):
+            else:
+                logging.info(f"Created queue folder {queue_dir}")
+                is_new_queue = True
+
+            # Move the request file to the queue folder.
+            try:
+                os.rename(f.name, request_path)
+            except OSError as e:
+                # If the queue folder was deleted in the meantime, try again.
+                if e.errno == errno.ENOENT:
+                    continue
+                else:
+                    raise e
+            else:
+                logging.info(f"Queued request {request_path}")
+            break
+    except Exception as e:
+        logging.error(f"Error while queuing request", exc_info=e)
+        os.remove(f.name)
+        raise e
+
+    return is_new_queue
 
-    headers = {"Content-Type": "application/json"}
-    request(callback_url, result, headers)
+
+def process_requests(url, unattested):
+    queue_dir = get_queue_dir(url)
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        tmp_queue_dir = os.path.join(tmp_dir, "queue")
+
+        try:
+            # Process the first request by doing an actual fetch.
+            # The queue folder is not moved yet, so that more requests
+            # for the same URL can be queued while the fetch is in progress.
+            request_path = os.path.join(queue_dir, os.listdir(queue_dir)[0])
+            logging.info(f"Processing first request {request_path}")
+            with open(request_path) as f:
+                request_metadata = json.load(f)
+
+            # Add nonce as query parameter for cache busting.
+            # This reduces the time to observe DID document updates
+            # for some servers like GitHub Pages.
+            url = url + f"?{request_metadata['nonce']}"
+
+            if unattested:
+                result = fetch_unattested(url, request_metadata["nonce"])
+            else:
+                result = fetch_attested(url, request_metadata["nonce"])
+
+            # Send back the result to the callback URL.
+            callback_data = {"result": json.loads(result)}
+            body = json.dumps(callback_data).encode("utf-8")
+            headers = {"Content-Type": "application/json"}
+            request(request_metadata["callback_url"], body, headers)
+
+            # Remove the request file to prevent it from being processed again below.
+            os.remove(request_path)
+
+        except Exception as e:
+            logging.error(
+                f"Error while processing first request in {queue_dir}", exc_info=e
+            )
+            logging.info(f"Removing queue {queue_dir} and aborting")
+            # Move the queue dir to a temporary location that will be deleted.
+            os.rename(queue_dir, tmp_queue_dir)
+            raise
+
+        # Move the queue folder to a temporary location that will be deleted and
+        # process any remaining requests that were added in the meantime.
+        os.rename(queue_dir, tmp_queue_dir)
```

Comment:
Make this a finally block of the above try/except, then you don't need to do it in two different places.

```suggestion
        except Exception as e:
            logging.error(
                f"Error while processing first request in {queue_dir}", exc_info=e
            )
            logging.info(f"Removing queue {queue_dir} and aborting")
            raise
        finally:
            # Move the queue folder to a temporary location that will be deleted and
            # process any remaining requests that were added in the meantime.
            os.rename(queue_dir, tmp_queue_dir)
```